buildSrc/src/main/kotlin/SystemTestPorts.kt

```diff
@@ -0,0 +1,69 @@
+import org.gradle.api.Project
+import org.gradle.api.provider.Provider
+import org.gradle.api.services.BuildService
+import org.gradle.api.services.BuildServiceParameters
+import org.gradle.api.tasks.testing.Test
+
+/**
+ * Synchronizes access to ports for the system tests so two system tests running in parallel don't
+ * accidentally try to use the same port.
+ */
+abstract class SystemTestPorts : BuildService<BuildServiceParameters.None> {
+
+	private var nextFreePort = 6000
+	private val lock = Object()
+
+	/**
+	 * Used ports should be unique for each system test to avoid collisions during parallel execution.
+	 * Thus, we assign a new port each time this is called.
+	 * Since tests may run in different workers that don't share the #nextFreePort variable, we use the worker number
+	 * to avoid collisions (given that we don't have more than 100 system tests per worker).
+	 */
+	fun pickFreePort(): Int {
+		synchronized(lock) {
+			val pickedPort = nextFreePort
+			nextFreePort += 1
+			return pickedPort
+		}
+	}
+
+	companion object {
+
+		/** The provider for the build service. You must call #registerWith before using this property. */
+		lateinit var provider: Provider<SystemTestPorts>
+
+		/** Registers the SystemTestPorts with the given project. */
+		fun registerWith(project: Project) {
+			provider =
+				project.gradle.sharedServices.registerIfAbsent("system-test-ports", SystemTestPorts::class.java) {
+					maxParallelUsages.set(1)
+				}
+		}
+
+		/** @see SystemTestPorts.pickFreePort */
+		fun pickFreePort(): Int {
+			return provider.get().pickFreePort()
+		}
+	}
+
+}
+
+private val teamscalePortMap = mutableMapOf<Test, Int>()
+
+/** The port that the fake Teamscale should use during the system test. Guaranteed conflict-free. */
+val Test.teamscalePort: Int
+	get() {
+		synchronized(teamscalePortMap) {
+			return teamscalePortMap.getOrPut(this) { SystemTestPorts.pickFreePort() }
```

Comment:
This might work, but I'm not really sure to be honest as we are now sharing one instance of the Provider potentially across projects. I would suggest the following:

Store the chosen provider and the chosen port as an extension property directly on the test instance so that we don't need to keep the test instances in a map, which kind of feels dangerous.
```
var Test.portProvider: Provider<SystemTestPorts>
	get() = extra["portProvider"] as Provider<SystemTestPorts>
	set(value) {
		extra["portProvider"] = value
	}

var Test.teamscalePort: Int
	get() = extra["teamscalePort"] as Int
	set(value) {
		extra["teamscalePort"] = value
	}
```

In `com.teamscale.system-test-convention.gradle.kts` you can do:
```
val provider = SystemTestPorts.registerWith(project)

tasks.test {
//...
	usesService(provider)
	portProvider = provider
	teamscalePort = provider.get().pickFreePort()
	agentPort = provider.get().pickFreePort()
	systemProperties("agentPort" to agentPort, "teamscalePort" to teamscalePort)
}
```
and hence don't have to rely on the global provider instance anymore.

For the redirect/artifactory ports it becomes `val artifactoryPort = portProvider.get().pickFreePort()`

I committed the changes to your branch as I already had them locally anyway. Please have look if you agree: https://github.com/cqse/teamscale-jacoco-agent/pull/287/commits/fa72d4daa0131ec810d24f8e98eabc918587170d

Reply:
looks excellent, thank you!